theano/scan_module/scan_opt.py

```diff
@@ -219,106 +228,123 @@ def apply(self, fgraph):
+                    if x in inner_non_seqs_set:
+                        _idx = inner_non_seqs_map[x]
+                        outside_ins.append(outer_non_seqs[_idx])
+                    elif x in to_replace_set:
+                        outside_ins.append(replace_with_out[to_replace_map[x]])
+                    elif isinstance(x, theano.Constant):
+                        outside_ins.append(x.clone())
+                    else:
+                        raise Exception(
+                            ('Error in the `scan_pushout_non_seq_'
+                             'operations`. The optimization tries '
+                             'to move some computation fron scan '
+                             'which is not allowed to move. Report '
+                             'this on theano-users list'), x)
+                outside_ins = [x.type.filter_variable(y) for x, y in
+                               zip(nd.inputs, outside_ins)]
+
+                # Do not call make_node for test_value
+                nw_outer_node = nd.op(*outside_ins,
+                                      **dict(return_list=True))[0].owner
+
+                # Step 2. Create variables for replacements
+                for idx, y in enumerate(nd.outputs):
+                    y_place_holder = scan_utils.safe_new(y, '_replace')
+                    nto_replace = add_to_replace(y, nto_replace)
+                    replace_with_in.append(y_place_holder)
+                    assert isinstance(y, type(nw_outer_node.outputs[idx]))
+                    replace_with_out.append(nw_outer_node.outputs[idx])
+
         # We need to check all candidate replacements and choose those that
         # make sense for us
-
         # Step 1. which elements of `to_replace` are used by remaining
         # components of the inner function
         clean_to_replace = []
         clean_replace_with_in = []
         clean_replace_with_out = []
-        existent_nodes = [nd for nd in local_fgraph.toposort()
-                          if nd not in to_remove]
-        to_keep = []
-        for nd in existent_nodes:
-            to_keep += nd.inputs
-        for idx, out in enumerate(to_replace):
-            if (out in to_keep
-                    and out.owner not in existent_nodes
-                    # If types are different, conversion Op will be inserted,
-                    # and it may trigger an infinite loop.
-                    and replace_with_in[idx].type == out.type):
-                clean_to_replace += [out]
-                clean_replace_with_in += [replace_with_in[idx]]
-                clean_replace_with_out += [replace_with_out[idx]]
+        existent_nodes = [nd for nd in local_fgraph_topo
+                          if nd not in to_remove_set]
+        existent_nodes_set = set(existent_nodes)
+        to_keep = []; [to_keep.extend(nd.inputs) for nd in existent_nodes]
+        to_keep_set = set(to_keep)
+
+        for out, idx in to_replace_map.items():
```

Comment:
I do not think there will be a big speed up and this is not inside the big loop, hence this is not the main bottleneck. Plus we will have to keep track of and maintain a one more additional list just for that.